eyeglass 0.8.0 support

JohnRiv opened the pull request:
Includes some dependency version updates to match eyeglass 0.8.0's dependencies. I also had to make a couple adjustments to support eyeglass 0.8.0 in a very basic test project I set up so that spriting worked again.

This closes Issue #18

JohnRiv committed:
eyeglass 0.8.0 support

Includes some dependency version updates to match eyeglass 0.8.0's dependencies

This closes Issue #18

alanhogan commented:
Nice.  I've been waiting to this and broccoli-eyeglass

> On Dec 22, 2015, at 11:33 PM, John Riviello notifications@github.com wrote:
> 
> Includes some dependency version updates to match eyeglass 0.8.0's dependencies. I also had to make a couple adjustments to support eyeglass 0.8.0 in a very basic test project I set up so that spriting worked again.
> 
> This closes Issue #18
> 
> You can view, comment on, or merge this pull request online at:
> 
>   https://github.com/sass-eyeglass/eyeglass-spriting/pull/19
> 
> Commit Summary
> 
> eyeglass 0.8.0 support
> File Changes
> 
> M eyeglass-exports.js (2)
> M package.json (12)
> M sass/index.scss (2)
> Patch Links:
> 
> https://github.com/sass-eyeglass/eyeglass-spriting/pull/19.patch
> https://github.com/sass-eyeglass/eyeglass-spriting/pull/19.diff
> —
> Reply to this email directly or view it on GitHub.

eoneill commented on package.json:
```diff
@@ -32,13 +32,13 @@
   "dependencies": {
     "lwip": "^0.0.8",
-    "node-sass-utils": "^1.1.1",
+    "node-sass": "^3.4.2",
```
Can you move `node-sass` back into `devDependencies`

This is something we're still working through on the module side, but we've seen weird issues in the past when a module and a build system both declare node-sass dependencies. It's better to just let the build system declare it's dependency.

We need it in `devDependencies` for testing.

JohnRiv replied:
I believe it actually has to be a dependency because Layout.js & SpriteMap.js both `require("node-sass")`. More details in [my comment below](https://github.com/sass-eyeglass/eyeglass-spriting/pull/19#issuecomment-167058841).

eoneill commented on sass/index.scss:
```diff
@@ -1,5 +1,3 @@
-@import "eyeglass/assets";
```
Can you comment on why this was removed?

JohnRiv replied:
Details regarding why are in [my comment below](https://github.com/sass-eyeglass/eyeglass-spriting/pull/19#issuecomment-167058841).

eoneill commented:
Can you also look over the [Travis CI job](https://travis-ci.org/sass-eyeglass/eyeglass-spriting/jobs/98462868) and fix the failing tests (don't worry about the `iojs` failure).

Some of the tests currently output the map result into CSS, which causes node-sass 3.4+ to throw a `... isn't a valid CSS value` exception.

You can fix this by changing the tests so they output the map into a comment rather than as a CSS property.

e.g.

``` scss
/* #{inspect(sprite-map(...))} */
```

Thanks for the PR and lemme know if you have any questions.

JohnRiv commented:
I'll see what I can do about the failures on Travis CI. In the meantime, I've pushed up my sample project and a separate branch of eyeglass-spriting that has a individual commit for each step so you can see why I made each change I did.

**Sample Project Commits:** https://github.com/JohnRiv/eyeglass-spriting-grunt/commits/master

**eyeglass-spriting individual commits that fix the errors along the way in the sample project:** https://github.com/JohnRiv/eyeglass-spriting/commits/eyeglass08steps

You can view those for more details, but the basic answers to your other 2 questions/requests are:
1. `node-sass` has to be a dependency of eyeglass-spriting instead of a devDependency because [Layout.js](https://github.com/sass-eyeglass/eyeglass-spriting/blob/21b60461cd701289ebcea2c42177e68251876a94/Layout.js#L9) & [SpriteMap.js](https://github.com/sass-eyeglass/eyeglass-spriting/blob/21b60461cd701289ebcea2c42177e68251876a94/SpriteMap.js#L12) both `require("node-sass")`
   - The `Error: Cannot find module 'node-sass'` error that is thrown before this change was made is documented in [this commit of my sample project](https://github.com/JohnRiv/eyeglass-spriting-grunt/commit/d3f1560863dd3794917232164de8c4579fd145ce)
2. The `@import "eyeglass/assets";` line was removed because simply adding `@import "spriting";` to my Sass file caused `Fatal error: Cannot read property 'eyeglass' of undefined` when compiling my Sass. I traced that down to the error being thrown in eyeglass's eyeglass_modules.js in the [`canAccessModule` function where it defines the `canAccess` variable](https://github.com/sass-eyeglass/eyeglass/blob/bdc5b2158fc0055aa082819e7eae92b1f1aa12ba/lib/util/eyeglass_modules.js#L275-L280).
   That has an [if statement that checks if branch.dependencies[name] is truthy](https://github.com/sass-eyeglass/eyeglass/blob/bdc5b2158fc0055aa082819e7eae92b1f1aa12ba/lib/util/eyeglass_modules.js#L278).
   With that `@import "eyeglass/assets";` line, `dependencies` is undefined
   because the `branches` variable is equal to:

``` js
[ { name: 'spriting',
    version: '0.9.2',
    path: '/projects/js/eyeglass-spriting-grunt/node_modules/eyeglass-spriting' } ]
```

The `name` variable in this case is `eyeglass`, hence the error message.
- The error that is thrown before this change was made is documented in [this commit of my sample project](https://github.com/JohnRiv/eyeglass-spriting-grunt/commit/67cb14e987797de757dd0c2a69c504b31e01202f)

JohnRiv commented:
node v0.12 passes now with my latest commit. Thanks for the tip @eoneill, it would have taken me forever to figure that out

JohnRiv commented:
Just had a thought about the require("node-sass") statements... Since eyeglass is a dependency and eyeglass has node-sass as one of its dependencies, we can update the require("node-sass") lines to reference eyeglass's node-sass via require("eyeglass/node_modules/node-sass"). Would that work better?

eoneill commented:
I'll try to take a look at this today or tomorrow. Thanks for your patience!

JohnRiv commented:
Thanks. I just realized today while working with a project using NPM 3 that `require("eyeglass/node_modules/node-sass")` won't work with NPM 3 since peerDependencies are no longer installed in a node_modules subfolder. So I think we're back to having to make node-sass a dependency for eyeglass-spriting, unless you have another suggested approach.

JohnRiv commented:
I haven't tested this yet & it's a bit ugly, but I think this would actually be an NPM2 & NPM3 compatible way to require node-sass if you can't list it as a dependency, given that node-sass is a dependency of Eyeglass and Eyeglass is a dependency of eyeglass-spriting:

``` js
var sass;
try {
  sass = require("node-sass"); // npm3
} catch (err) {
  sass = require("eyeglass/node_modules/node-sass"); // npm2
}
```

eoneill commented on eyeglass-exports.js:
```diff
@@ -143,7 +143,7 @@ module.exports = function(eyeglass, sass) {
 
           // create the image in the eyeglass cache
-          var spritemapsDir = path.join(eyeglass.options.cacheDir, "spritemaps");
+          var spritemapsDir = path.join(eyeglass.options.eyeglass.cacheDir, "spritemaps");
```
for back-compat reasons, can you make this...

``` js
var options = eyeglass.options;
var cacheDir = options.eyeglass && options.eyeglass.cacheDir || options.cacheDir;
var spritemapsDir = path.join(cacheDir, "spritemaps");
```

eoneill commented on test/test_spriting.js:
```diff
@@ -13,9 +13,9 @@ describe("spriting module", function () {
-                ".sprite-map-test { foo: sprite-map('test-sprite-map', sprite-layout(horizontal, " +
-                "(spacing: 5px, alignment: bottom)), 'images/*'); }";
-    var expected = ".sprite-map-test {\n  foo: (sprite-map: true, name: test-sprite-map, sources:" +
+                "/* #{inspect(sprite-map('test-sprite-map', sprite-layout(horizontal, " +
```
You can revert these changes and rebase. I merged my PR (#17) that fixes these test (which I forgot I had created).

eoneill commented:
Sorry for the delay.

Regarding the `require("node-sass")` issue, I think we'll need to refactor things a bit so they're not depending on node-sass in this way.

They way it should be designed is that the custom functions have access to the eyeglass and sass instance being used, so they should get passed into the necessary places (e.g. Layout.js & SpriteMap.js). Let me know if you want to take this on or not.

Regarding the `@import "eyeglass/assets"`, I think this is a bug in eyeglass which I will take a look at.

eoneill commented:
Re `@import "eyeglass/assets"`, I think this will fix the issue: https://github.com/sass-eyeglass/eyeglass/pull/118

eoneill committed:
remove require(node-sass) in favor of sass passed via custom function invocation

eoneill commented:
Here's an example of what I mean by removing the `require("node-sass")` all together and relying on the version of `sass` passed in the custom functions invocation:
https://github.com/eoneill/eyeglass-spriting/commit/a6a39a4a131773afab279608b382fa1f40eebe15

JohnRiv committed:
Backwards compatible cacheDir

JohnRiv committed:
Moving node-sass back to devDependencies

JohnRiv commented:
Thanks. I removed my test updates in favor of your commit to master that I rebased, put in the backwards compatible cacheDir, moved `node-sass` from `dependencies` back to `devDependencies`, and it seems your eoneill@a6a39a4 covered all the `require("node-sass")` instances so I just cherry-picked your change on to this. I tested it out and it worked, so I've pushed up those changes. Node.js 0.12 tests passed on TravisCI.

I believe at this point all we need is sass-eyeglass/eyeglass#118 to be merged into Eyeglass and then after Eyeglass 0.8.2 is released with that change, I can remove the `@import "eyeglass/assets";` reference, update the Eyeglass dependency to 0.8.2, and we'll be all set.

Let me know if I missed anything

JohnRiv commented:
By "remove the `@import "eyeglass/assets";` reference" I of course meant to put it back in :wink: 

I have the changes staged locally to test out as soon as Eyeglass 0.8.2 is released, let me know when that is available and I'll test my changes and push them up so we can close out this pull request and have a 0.8.x version of eyeglass-spriting available.

eoneill commented:
`eyeglass@0.8.2` has been released

JohnRiv committed:
Eyeglass 0.8.2 support

Adding back `@import "eyeglass/assets"` now that sass-eyeglass/eyeglass#118 is merged

Updated the Eyeglass dependency to 0.8.2

JohnRiv commented:
Thanks for the heads up. I just pushed up another commit that uses Eyeglass ^0.8.2 and added back `@import "eyeglass/assets";`. I tested it out on a couple projects of mine and it worked great. Tests are passing on Node 0.12 as well. Let me know if there's anything else you'd like me to do.

JohnRiv commented:
Thanks for merging. When do you anticipate cutting a release with this and the other improvements you've made?
